### PR TITLE
Thread-safe dispatcher table

### PR DESCRIPTION
**Pull request by ilia-cher**
Summary: Allow concurrent read and writes in dispatcher table

Differential Revision: D8722560

### CODE REVIEW

File: caffe2/core/dispatch/DispatchTable.h [hunk 1]
```diff
@@ -1,104 +1,118 @@
...
+    for (auto* map : maps_to_delete_) {
+      delete map;
     }
+  }
 
-    void erase(const Key& key) {
-      // TODO Locking
-      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+  template <class Key_>
+  void emplace(Key_&& key, void* value) {
```
Comment (ezyang) [inline caffe2/core/dispatch/DispatchTable.h]: Shouldn't we encourage bulk inserts? Otherwise, O(n) emplaces will cost O(n^2).

Comment (ilia-cher) [reply caffe2/core/dispatch/DispatchTable.h]: good suggestion, but I think this is outside of the scope for this diff

Comment (smessmer) [reply caffe2/core/dispatch/DispatchTable.h]: How does it cost `O(n^2)`? The hashmap will only reallocate when crossing a `2^n` size boundary, so inserting should be amortized `O(n)`.

Comment (smessmer) [reply caffe2/core/dispatch/DispatchTable.h]: The current implementation is overly expensive. We should avoid allocating a new hash map every time, then we'd be amortized `O(n)`. It should be possible to just keep two hash maps around, which are both always identical. One is read from, writing writes to the other one, swaps them, then writes the same change to the first one.

Comment (ezyang) [reply caffe2/core/dispatch/DispatchTable.h]: Well, it didn't use to be O(n^2), because we weren't copying the hashmap before?

**Pull request closed without merging**